File: scaffolds/leetcode/1636_sort_array_by_increasing_frequency/1636_sort_array_by_increasing_frequency.go
func frequencySort(nums []int) []int {
	freq := make(map[int]int)
	for _, num := range nums {
		freq[num]++
	}

	type pair struct {
		value int
		count int
	}
	pairs := make([]pair, len(nums))
	for i, num := range nums {
		pairs[i] = pair{value: num, count: freq[num]}
	}

	sort.Slice(pairs, func(i, j int) bool {
		if pairs[i].count == pairs[j].count {
			return pairs[i].value > pairs[j].value
		}
		return pairs[i].count < pairs[j].count
	})

	for i, p := range pairs {
		nums[i] = p.value
	}
	return nums
}

File: scaffolds/leetcode/1636_sort_array_by_increasing_frequency/1636_sort_array_by_increasing_frequency.java
class Solution {
    public int[] frequencySort(int[] nums) {
        Map<Integer, Integer> freq = new HashMap<>();
        Integer[] newNums = new Integer[nums.length];

        for (int i = 0; i < nums.length; i++) {
            freq.put(nums[i], freq.getOrDefault(nums[i], 0) + 1);
            newNums[i] = nums[i];
        }

        Arrays.sort(newNums, (n1, n2) -> {
            if (freq.get(n1) != freq.get(n2)) {
                // if freq of numbers are not equal then return in increasing order based on
                // freq.
                return freq.get(n1) - freq.get(n2);
            } else {
                // otherwise sort them in decreasing order based on number in nums.
                return n2 - n1;
            }
        });

        for (int i = 0; i < nums.length; i++) {
            nums[i] = newNums[i];
        }

        return nums;
    }
}

class Solution {
	public:
		vector<int> frequencySort(vector<int>& nums) {
			vector<int> count(201, 0);
			for (int num : nums) {
				count[num + 100]++;
			}
			sort(nums.begin(), nums.end(), [&](int a, int b) {
				if (count[a + 100] == count[b + 100])
					return a > b;
				return count[a + 100] < count[b + 100];
			});
			return nums;
		}
	};

